### DISCUSSION TITLE
关于StyledComponents和Hooks的疑问

### DISCUSSION THREAD

**Discussion by nonzzz** [zxuqian/frontend-questions#8] [求助]
## 描述一下问题？

![~ORK~ILCYF{2JWX9IS%I{ I](https://user-images.githubusercontent.com/52351095/98221517-b22a5a00-1f8a-11eb-9251-0cf08bdc1306.png)
上图代码中active控制 opacity 显示隐藏我是在 styled里面写好呢。还是在jsx文件里面写。通过style属性。 然后我有个类似Vue computed的计算动态添加style属性我是写在styled里还是 用useHooks呢。我看了文档。官方文档关于对useHooks的描述。在函数组件主体内（这里指在 React 渲染阶段）改变 DOM、添加订阅、设置定时器、记录日志以及执行其他包含副作用的操作都是不被允许的，因为这可能会产生莫名其妙的 bug 并破坏 UI 的一致性。因此对这个钩子感到疑惑。我的认知是这个钩子一般用来处理http请求 。或者其他副作用操作。求群主解惑。

**Comment by zxuqian** [comment]
1. active 推荐直接传给 styled-components。
2. 如果需要在 JSX 里判断，直接写在函数体里就行，React 不分 computed 还是普通属性，只有属性和状态的区别，状态是会引起组件刷新的，需要使用 useState()定义。属性不会引起刷新，通过组件参数传递就可以了。
3. hooks 只是一段代码，任何函数体里的代码都可以放到自定义的 hooks 中，目的是为了复用。
4. 改变 DOM、添加订阅、设置定时器、记录日志以及执行其他包含副作用的操作，这些使用内置的 useEffect() 或 useRef() hooks，不能在函数体里直接写，否则不会引起组件刷新。

**Comment by zxuqian** [comment]
如果你用 className 的话，那么就根据active 属性的值，去计算该用的 className
用 style 的话，可以直接用 object 传递相关的 CSS 样式
用 styled-components 的话，直接用属性给包装后的组件传值就行
至于属性和组件缓存，还有active 的计算逻辑，可以直接写在函数体里
如果在其它组件中需要复用这个逻辑，或者为了减少函数里的代码，可以抽成 hooks

**Comment by nonzzz** [comment]
![{6KQ1F}5U _@N`N{0U~IP_X](https://user-images.githubusercontent.com/52351095/98440263-dde84400-2132-11eb-87dc-618df2b7d18c.png)
群主想问下。为什么我采用<Route>包裹组件</Route>拿不到 路由传递进来的参数呢。写成component 就行。看了下router的文档，也没看出个所以然。。

**Comment by zxuqian** [comment]
使用 children 这种形式需要使用 hooks 来访问路由参数：

```jsx
  const location = useLocation();
  const history = useHistory();
  const match = useRouteMatch();
  const params = useParams();
```
[示例：CodeSandBox](https://codesandbox.io/s/react-router-children-get-router-params-nkyfb?fontsize=14&hidenavigation=1&theme=dark)
